README.md
node-yaowst
===========

Yet another OpsWorks ssh Tool


CHANGELOG.markdown
# [Yet another OpsWorks ssh Tool (YaOWsT)](./README.markdown) ChangeLog

## 2018-09-11, [v0.1.1](https://github.com/w4andy/node-yaowst/tree/v0.1.1)

### Commits

  - [[`7f3f9d6724`](https://github.com/w4andy/node-yaowst/commit/7f3f9d6724b2b81ce1a2a706754e070b77287ce6)] - **CHANGELOG**: fix commit urls
  - [[`656b0451f9`](https://github.com/w4andy/node-yaowst/commit/656b0451f9de8e6735a9ef2f73aa4a588792f839)] - **deps**: aws-sdk v2.332.0, cli-spinner v0.2.8, deep-extend v0.6.0, shelljs v0.8.2, step v1.0.0, yargs v12.0.2, eslint v5.6.1, mocha v5.2.0, next-update v3.6.0, nyc v13.0.1
  - [[`a3e9953812`](https://github.com/w4andy/node-yaowst/commit/a3e99538127b5bbb1d4cded282f4507ddb04b2f7)] - **doc**: update email address
  - [[`ce212489e3`](https://github.com/w4andy/node-yaowst/commit/ce212489e3c39c3fe41e719ac4c9eacfecd9920b)] - **doc**: fix changelog
  - [[`9c62c13d7f`](https://github.com/w4andy/node-yaowst/commit/9c62c13d7f75f132ebc6b6020ff77974155db955)] - **test**: fix sshconfig suiteTeardown
  - [[`14032cedaa`](https://github.com/w4andy/node-yaowst/commit/14032cedaa9a7401de65691af0673542923e5810)] - fix release date

## 2016-08-28, [v0.1.0](https://github.com/w4andy/node-yaowst/tree/v0.1.0)

### Notable changes

 - add OpsWorks region to the config

### Commits

  - [[`4f48370d46`](https://github.com/w4andy/node-yaowst/commit/4f48370d46297dffab662844ba3c5b161aeb9681)] - **deps**: downgrade next-update to 1.3, 1.5 broke the npm install on ubuntu trusty 64bit with node v6
  - [[`6af8839884`](https://github.com/w4andy/node-yaowst/commit/6af8839884eafb85be884e3f4fafc98efb606877)] - **deps**: update deps aws-sdk, cli-spinner, deep-extend, shelljs, yargs, changelog42, mocha, next-update
  - [[`aa22f35e12`](https://github.com/w4andy/node-yaowst/commit/aa22f35e12f3e910812cdff5115d0385e7163e7d)] - **docu**: add uninstall information
  - [[`5667ef9d6b`](https://github.com/w4andy/node-yaowst/commit/5667ef9d6b9966f6e93fdc6f6967f660b63b983e)] - **lib**: add OpsWorks region to the config
  - [[`5ffbd328cd`](https://github.com/w4andy/node-yaowst/commit/5ffbd328cd6c9cabe0855eaa4416a723a3a00f46)] - **lib/sshconfig.js**: add double quote to the host
  - [[`94dc9ab1ec`](https://github.com/w4andy/node-yaowst/commit/94dc9ab1ec61dfa85210bb6c5a9f1fdd7373e7ed)] - **test**: add nyc and add node v6 to the .travis.yml
  - [[`49c2efc4b6`](https://github.com/w4andy/node-yaowst/commit/49c2efc4b6ea25fb43d89e54c01b169a7c1bca01)] - **test**: use eslint instead of jshint
  - [[`81f7bd0f1b`](https://github.com/w4andy/node-yaowst/commit/81f7bd0f1b676abed7bd4bdbfa8586c2d2acbfac)] - **test** update node version


## 2015-11-22, [v0.0.2](https://github.com/w4andy/node-yaowst/tree/v0.0.2)

### Commits

  - [[`4bc9546118`](https://github.com/w4andy/node-yaowst/commit/4bc95461187b88fa6c523490619f5d12497ea241)] - **doc**: add more hints to find help
  - [[`43d8e45495`](https://github.com/w4andy/node-yaowst/commit/43d8e4549505c832b228ea1e183af1289575d7c0)] - **doc**: Add "Example Config"
  - [[`146aede377`](https://github.com/w4andy/node-yaowst/commit/146aede377b0775580a69943c1c38333cc59df29)] - **lib/sshconfig.js**: fixed adding new line after writing the post block


## 2015-11-15, [v0.0.1](https://github.com/w4andy/node-yaowst/tree/v0.0.1)

First release.

---

_The commit log is generated with [changelog42](https://www.npmjs.com/package/changelog42)._

test/config.js
'use strict';

/* global suite: false, setup: false, test: false,
 teardown: false, suiteSetup: false, suiteTeardown: false */

var assert = require('assert');
var os = require('os');
var fs = require('fs');
var path = require('path');
var Config = require('../lib/config');

suite('Config', function() {
  var filePathWriteTest = path.join(os.tmpDir(), 'configWriteTest_' + new Date().getTime() + '.json'),
    filePathReadTest = path.join(__dirname, '../resources/test/configReadTest.json');

  suiteSetup(function(done) {
    // check if read test file has 0600
    fs.exists(filePathReadTest, function(exists) {
      if (exists) {
        fs.stat(filePathReadTest, function(err, fileStat) {
          if (err) {
            done(err);
          } else {
            var fileMode = parseInt(fileStat.mode.toString(8), 10).toString().substr(-3);
            if (fileMode === '600' || fileMode === '400') {
              done();
            } else {
              fs.chmod(filePathReadTest, '0600', function(err2) {
                done(err2);
              });
            }
          }
        });
      } else {
        done();
      }
    });
  });


  test('write the base_config.json file', function(done) {
    var config = new Config(filePathWriteTest),
      baseConfig = {
        opsWorks: {
          region: 'us-east-1'
        },
        sshOptions: {
          StrictHostKeyChecking: 'no',
          UserKnownHostsFile: '/dev/null',
          IdentitiesOnly: 'yes'
        },
        stacks: []
      };

    config.writeDefaultFile(true, function(err) {
      if (err) {
        done(err);
      } else {
        var fileContent = fs.readFileSync(filePathWriteTest, 'utf8').trim(),
          fileContentJson = JSON.parse(fileContent);

        assert.strictEqual(JSON.stringify(fileContentJson), JSON.stringify(baseConfig));
        done();
      }
    });
  });

  test('read the config file', function(done) {
    var config = new Config(filePathReadTest),
      expectedData = {
        stacks: [
          {
            name: 'opsworks_one',
            id: 'sssssss',
            sshOptions: {
              StrictHostKeyChecking: 'yes',
              IdentityFile: '~/.ssh/opsworks'
            },
            layers: [
              {
                id: 'd1',
                name: 'xx',
                alias: 'opsworks_one_one_'
              },
              {
                id: 'd2',
                name: 'yy',
                alias: 'opsworks_one_two_',
                sshOptions: {
                  Port: 9999
                }
              },
              {
                id: 'd3',
                name: 'zz',
                alias: 'opsworks_one_two_',
                sshOptions: {
                  User: 'root'
                }
              }
            ]
          }
        ],
        sshOptions: {
          StrictHostKeyChecking: 'no',
          UserKnownHostsFile: '~/.ssh/known_hosts_opsworks',
          IdentitiesOnly: 'yes',
          User: 'node'
        },
        sshConfigFile: {
          file: null,
          saveMode: '24/7'
        },
        opsWorks: {
          region: 'us-east-1',
          accessKeyId: 'xxxxxxxxx',
          secretAccessKey: 'yyyyyyyy'
        }
      };

    config.getConfig(function(err, configData) {
      if (err) {
        done(err);
      } else {
        assert.strictEqual(JSON.stringify(configData), JSON.stringify(expectedData));
        done();
      }
    });
  });

  test('read the config file and enrichment the config data', function(done) {
    var config = new Config(filePathReadTest),
      expectedData = {
        stacks: [
          {
            name: 'opsworks_one',
            id: 'sssssss',
            sshOptions: {
              User: 'node',
              IdentitiesOnly: 'yes',
              UserKnownHostsFile: '~/.ssh/known_hosts_opsworks',
              StrictHostKeyChecking: 'yes',
              IdentityFile: '~/.ssh/opsworks'
            },
            layers: [
              {
                id: 'd1',
                name: 'xx',
                alias: 'opsworks_one_one_'
              },
              {
                id: 'd2',
                name: 'yy',
                alias: 'opsworks_one_two_',
                sshOptions: {
                  Port: 9999
                }
              },
              {
                id: 'd3',
                name: 'zz',
                alias: 'opsworks_one_two_',
                sshOptions: {
                  User: 'root'
                }
              }
            ],
            accessKeyId: 'xxxxxxxxx',
            secretAccessKey: 'yyyyyyyy',
            region: 'us-east-1'
          }
        ],
        sshOptions: {
          StrictHostKeyChecking: 'no',
          UserKnownHostsFile: '~/.ssh/known_hosts_opsworks',
          IdentitiesOnly: 'yes',
          User: 'node'
        },
        sshConfigFile: {
          file: null,
          saveMode: '24/7'
        },
        opsWorks: {
          region: 'us-east-1',
          accessKeyId: 'xxxxxxxxx',
          secretAccessKey: 'yyyyyyyy'
        }
      };

    config.getEnhancedConfig(function(err, configData) {
      if (err) {
        done(err);
      } else {
        assert.strictEqual(JSON.stringify(configData), JSON.stringify(expectedData));
        done();
      }
    });
  });


  suiteTeardown(function() {
    fs.unlinkSync(filePathWriteTest);
  });
});


test/yaowst.js
'use strict';

/* global suite: false, setup: false, test: false,
 teardown: false, suiteSetup: false, suiteTeardown: false */

var assert = require('assert');
var os = require('os');
var fs = require('fs');
var path = require('path');
var Yaowst = require('../lib/yaowst');

suite('Yaowst', function() {
  this.timeout(5000);
  var configFilePathWrite = path.join(__dirname, '../resources/test/configYaowst.json');

  suiteSetup(function(done) {
    // check if test config file has 0600
    fs.exists(configFilePathWrite, function(exists) {
      if (exists) {
        fs.stat(configFilePathWrite, function(err, fileStat) {
          if (err) {
            done(err);
          } else {
            var fileMode = parseInt(fileStat.mode.toString(8), 10).toString().substr(-3);
            if (fileMode === '600' || fileMode === '400') {
              done();
            } else {
              fs.chmod(configFilePathWrite, '0600', function(err2) {
                done(err2);
              });
            }
          }
        });
      } else {
        done();
      }
    });
  });

  test('first init yaowst', function(done) {
    var configFilePath = path.join(os.tmpDir(), 'configWriteTest_' + new Date().getTime() + '.json'),
      sshConfigFile = path.join(os.tmpDir(), 'sshConf_' + new Date().getTime());
    var yaowst = new Yaowst({
      configFile: configFilePath,
      sshConfigFile: {file: sshConfigFile}
    }, function(err) {
      assert.strictEqual(err.message, 'config file dose not exists!');
      yaowst.firstInit({}, function(err2) {
        if (err2) {
          done(err2);
        } else {
          fs.exists(configFilePath, function(exists) {
            assert.strictEqual(exists, true);
            fs.readFile(configFilePath, 'utf8', function(err3, fileContent) {
              if (err3) {
                done(err3);
              } else {
                var baseConfig = {
                  opsWorks: {
                    region: 'us-east-1'
                  },
                  sshOptions: {
                    StrictHostKeyChecking: 'no',
                    UserKnownHostsFile: '/dev/null',
                    IdentitiesOnly: 'yes'
                  },
                  stacks: []
                };
                var fileContentJson = JSON.parse(fileContent);
                assert.strictEqual(JSON.stringify(fileContentJson), JSON.stringify(baseConfig));
                done();
              }
            });
          });
        }
      });
    });
  });

  test('write ssh config', function(done) {
    var sshConfigFile = path.join(os.tmpDir(), 'sshConf_' + new Date().getTime());

    fs.exists(configFilePathWrite, function(exists) {
      if (exists) {
        var yaowst = new Yaowst({
          configFile: configFilePathWrite,
          sshConfigFile: {file: sshConfigFile}
        }, function(err) {
          if (err) {
            done(err);
          } else {
            yaowst.save({}, function(err2, cnt) {
              if (err2) {
                done(err2);
              } else {
                assert.strictEqual(cnt, 4);
                fs.exists(sshConfigFile, function(exists2) {
                  assert.strictEqual(exists2, true);
                  fs.readFile(sshConfigFile, 'utf8', function(err3, actualData) {
                    if (err3) {
                      done(err3);
                    } else {
                      actualData = actualData.replace(/HostName (?:[0-9]{1,3}\.){3}[0-9]{1,3}/g, 'HostName none');
                      var expectedData = '';
                      expectedData += '## yaowst begin ##\n';
                      expectedData += 'Host "opsworks_one_one_1"\n';
                      expectedData += '    HostName none\n';
                      expectedData += '    IdentitiesOnly yes\n';
                      expectedData += '    UserKnownHostsFile ~/.ssh/known_hosts_opsworks\n';
                      expectedData += '    StrictHostKeyChecking no\n';
                      expectedData += '    IdentityFile ~/.ssh/opsworks\n';
                      expectedData += '    User yaowst_test\n';
                      expectedData += '\n';
                      expectedData += 'Host "opsworks_one_two_*"\n';
                      expectedData += '    IdentitiesOnly yes\n';
                      expectedData += '    UserKnownHostsFile ~/.ssh/known_hosts_opsworks\n';
                      expectedData += '    StrictHostKeyChecking no\n';
                      expectedData += '    IdentityFile ~/.ssh/opsworks\n';
                      expectedData += '    User yaowst_test\n';
                      expectedData += '\n';
                      expectedData += 'Host "opsworks_one_two_1"\n';
                      expectedData += '    HostName none\n';
                      expectedData += '\n';
                      expectedData += 'Host "opsworks_one_two_2"\n';
                      expectedData += '    HostName none\n';
                      expectedData += '\n';
                      expectedData += '## yaowst end ##\n';

                      assert.strictEqual(actualData, expectedData);
                      done();
                    }
                  });
                });
              }
            });
          }
        });
      } else {
        // skip test if file not exists
        done();
      }
    });

  });

  suiteTeardown(function() {
  });
});


test/opsworks.js
'use strict';

/* global suite: false, setup: false, test: false,
 teardown: false, suiteSetup: false, suiteTeardown: false */

var assert = require('assert');
var fs = require('fs');
var path = require('path');
var OpsWorks = require('../lib/opsworks');

suite('OpsWorks', function() {
  this.timeout(5000);

  suiteSetup(function() {

  });

  test('get data per layer', function(done) {
    var configFile = path.join(__dirname, '../resources/test/configOpsWorksLayer.json');
    fs.exists(configFile, function(exists) {
      if (exists) {
        var opsWorksConfig = require(configFile);
        OpsWorks.fetchStackData(opsWorksConfig.stacks, function(err, data) {
          if (err) {
            done(err);
          } else {
            var expectedData = [
              {
                Host: 'opsworks_one_one_1',
                HostName: null,
                X_AutoScalingType: '24/7',
                User: 'yaowst_test',
                StrictHostKeyChecking: 'no',
                IdentityFile: '~/.ssh/opsworks',
                UserKnownHostsFile: '~/.ssh/known_hosts_opsworks'
              },
              {
                Host: 'opsworks_one_two_*',
                Port: 9999,
                User: 'yaowst_test',
                StrictHostKeyChecking: 'no',
                IdentityFile: '~/.ssh/opsworks',
                UserKnownHostsFile: '~/.ssh/known_hosts_opsworks',
                X_AutoScalingType: 'BASE_CONFIG'
              },
              {
                Host: 'opsworks_one_two_1',
                HostName: null,
                X_AutoScalingType: '24/7'
              },
              {
                Host: 'opsworks_one_two_2',
                HostName: null,
                X_AutoScalingType: '24/7'
              }
            ];

            expectedData.sort();

            // set ips to null
            var changedData = [];
            data.forEach(function(row) {
              if (row.HostName) {
                row.HostName = null;
              }
              changedData.push(row);
            });

            changedData.sort();

            assert.strictEqual(JSON.stringify(changedData), JSON.stringify(expectedData));
            done();
          }
        });
      } else {
        done();
      }
    });
  });

  test('get data per stack', function(done) {
    var configFile = path.join(__dirname, '../resources/test/configOpsWorksStack.json');
    fs.exists(configFile, function(exists) {
      if (exists) {
        var opsWorksConfig = require(configFile);
        OpsWorks.fetchStackData(opsWorksConfig.stacks, function(err, data) {
          if (err) {
            done(err);
          } else {
            var expectedData = [
              {
                Host: 'yaowst-*',
                UserKnownHostsFile: '~/.ssh/known_hosts_opsworks',
                IdentityFile: '~/.ssh/opsworks',
                StrictHostKeyChecking: 'no',
                User: 'yaowst_test',
                X_AutoScalingType: 'BASE_CONFIG'
              },
              {
                Host: 'yaowst-test-layer-one1',
                HostName: null,
                X_AutoScalingType: '24/7'
              },
              {
                Host: 'yaowst-test-layer-one2',
                HostName: null,
                X_AutoScalingType: '24/7'
              },
              {
                Host: 'yaowst-web1',
                HostName: null,
                X_AutoScalingType: '24/7'
              }];

            expectedData.sort();

            // set ips to null
            var changedData = [];
            data.forEach(function(row) {
              if (row.HostName) {
                row.HostName = null;
              }
              changedData.push(row);
            });

            changedData.sort();

            assert.strictEqual(JSON.stringify(changedData), JSON.stringify(expectedData));
            done();
          }
        });
      } else {
        done();
      }
    });
  });

  suiteTeardown(function() {

  });
});


lib/yaowst.js
'use strict';

const Config = require('./config');
const SshConfig = require('./sshconfig');
const opsWorks = require('./opsworks');

class Yaowst {

  constructor(options, callback) {
    /**
     * init options
     *
     * @type {object}
     * @private
     */
    this._options = options || {};

    /**
     * the yaowst config
     *
     * @type {object}
     * @private
     */
    this._config = {};

    /**
     * the ssh config object
     *
     * @type {SshConfig}
     * @private
     */
    this._sshConfigObject = null;

    // init config and load data
    /**
     * the config object
     *
     * @type {Config}
     * @private
     */
    this._configObject = new Config(options.configFile);
    this._configObject.getEnhancedConfig((err, config) => {
      if (err) {
        return callback(err);
      }
      if (!config.sshConfigFile) {
        config.sshConfigFile = {};
      }
      // overwrite ssh config file
      if (options.sshConfigFile) {
        if (options.sshConfigFile.file) {
          config.sshConfigFile.file = options.sshConfigFile.file;
        }
        if (options.sshConfigFile.saveMode) {
          config.sshConfigFile.saveMode = options.sshConfigFile.saveMode;
        }
      }
      this._config = config;
      callback();
    });
  }

  /**
   * get the hosts from ops works
   *
   * @param {function(?error, ?Array)} callback
   * @private
   */
  _loadHosts(callback) {
    opsWorks.fetchStackData(this._config.stacks, callback);
  }

  /**
   * get the ssh config object
   *
   * @return {SshConfig}
   * @private
   */
  _getSshConfigObject() {
    if (!this._sshConfigObject) {
      this._sshConfigObject = new SshConfig(this._options.sshConfigFile);
    }
    return this._sshConfigObject;
  }

  /**
   * save action
   * store the hosts into the ssh config file
   *
   * @param {object} option
   * @param {function(?error, ?number)} callback
   */
  save(option, callback) {
    let saveMode = this._config.sshConfigFile.saveMode;
    const sshConfigObject = this._getSshConfigObject();
    const allowedSaveModes = [sshConfigObject.SAVE_MODE_24_7, sshConfigObject.SAVE_MODE_ALL];

    if (option.saveMode && allowedSaveModes.indexOf(option.saveMode) === -1) {
      option.saveMode = null;
    }

    if (option.saveMode) {
      saveMode = option.saveMode;
    }

    this._loadHosts(function(err, hosts) {
      if (err) {
        return callback(err);
      }
      if (saveMode === sshConfigObject.SAVE_MODE_ALL) {
        sshConfigObject.writeConfig(hosts, (err2) => {
          callback(err2, hosts.length);
        });
      } else {
        let hostsToSave = [];
        const iterHosts = (cnt, cb) => {
          if (hosts.length > cnt) {
            const entry = hosts[cnt];

            if (entry.X_AutoScalingType === '24/7' || entry.X_AutoScalingType === 'BASE_CONFIG') {
              hostsToSave.push(entry);
            }

            setImmediate(() => {
              iterHosts(++cnt, cb);
            });
          } else {
            return cb();
          }
        };

        iterHosts(0, function(err3) {
          if (err3) {
            return callback(err3);
          }
          sshConfigObject.writeConfig(hostsToSave, (err4) => {
            callback(err4, hosts.length);
          });
        });
      }
    });
  }

  /**
   * first init action
   * create the config file and create a backup from the config file
   *
   * @param {object} option
   * @param {function(?error)} callback
   */
  firstInit(option, callback) {
    this._configObject.writeDefaultFile(option.force, (err) => {
      if (err) {
        return callback(err);
      }
      this._getSshConfigObject().backUpConfig(true, (err2) => {
        callback(err2);
      });
    });
  }
}


module.exports = Yaowst;


lib/sshconfig.js
'use strict';

const fs = require('fs');

/**
 * The ssh config handler
 *
 * @param {object|null} options
 * @constructor
 */
class SshConfig {

  constructor(options) {
    if (!options) {
      options = {};
    }
    if (!options.file) {
      options.file = process.env.HOME + '/.ssh/config';
    }
    Object.defineProperties(this, {
      _configFile: {
        value: options.file,
        writable: false
      },
      _stringConfig: {
        value: {
          ante: [],
          post: [],
          yaowst: []
        },
        writable: true
      },
      _config: {
        value: [],
        writable: true
      },
      CONFIG_BEGIN: {
        value: '## yaowst begin ##',
        writable: false
      },
      CONFIG_END: {
        value: '## yaowst end ##',
        writable: false
      },
      PART_TYPE_ANTE: {
        value: 'ante',
        writable: false
      },
      PART_TYPE_POST: {
        value: 'post',
        writable: false
      },
      PART_TYPE_YAOWST: {
        value: 'yaowst',
        writable: false
      },
      SAVE_MODE_ALL: {
        value: 'all',
        writable: false
      },
      SAVE_MODE_24_7: {
        value: '24/7',
        writable: false
      }
    });
  }

  /**
   * read the ssh config file and split the config data into ante yaowst, post yaowst and yaowst config lines
   *
   * @param {function(?error, ?object)} callback
   * @private
   */
  _readConfigFile(callback) {
    fs.stat(this._configFile, (err) => {
      if (err && err.code !== 'ENOENT') {
        return callback(err);
      } else if (err && err.code === 'ENOENT') {
        return callback(null, this._stringConfig);
      }
      fs.readFile(this._configFile, 'utf8', (readError, data) => {
        if (readError) {
          return callback(readError);
        } else {
          const lines = data.split('\n');
          let partType = this.PART_TYPE_ANTE;

          const iterLines = (cnt, cb) => {
            if (lines.length > cnt) {
              const line = lines[cnt].replace(/\s+$/, '');

              if (line === this.CONFIG_BEGIN) {
                partType = this.PART_TYPE_YAOWST;
              } else if (line === this.CONFIG_END) {
                partType = this.PART_TYPE_POST;
              } else {
                this._stringConfig[partType].push(line);
              }
              iterLines(++cnt, cb);
            } else {
              return cb();
            }
          };

          iterLines(0, (lineError) => {
            callback(lineError, this._stringConfig);
          });
        }
      });
    });
  }

  /**
   * backup the sshConfig file
   *
   * @param {Boolean} [initBackup] if true then use <filename>_backup_<date>
   * @param {function(?error)} callback
   */
  backUpConfig(initBackup, callback) {
    if (typeof initBackup === 'function') {
      callback = initBackup;
      initBackup = false;
    }

    fs.stat(this._configFile, (err, fileStat) => {
      if (err && err.code !== 'ENOENT') {
        return callback(err);
      } else if (err && err.code === 'ENOENT') {
        return callback();
      }
      fs.readFile(this._configFile, {encoding: 'utf8'}, (err2, data) => {
        if (err2) {
          return callback(err2);
        }
        const mode = parseInt(fileStat.mode.toString(8), 10).toString().substr(-3);
        let newFile = this._configFile + '.old';

        if (initBackup) { // use timestamp for init backup
          newFile = this._configFile + '.backup_' + new Date().toISOString().replace('T', '_').replace(/\..+/, '').replace(/:/g, '-');
        }

        fs.writeFile(newFile, data, {encoding: 'utf8', mode: mode}, function(err3) {
          callback(err3);
        });
      });
    });
  }


  /**
   * parse the yaowst config part
   *
   * @param {function(?error)} callback
   */
  parseConfig(callback) {
    this._readConfigFile((err) => {
      if (err) {
        return callback(err);
      }
      let hostConfig = {};
      const iterConfig = (cnt, cb) => {
        if (this._stringConfig.yaowst.length > cnt) {
          let line = this._stringConfig.yaowst[cnt];
          if (line.trim().substr(0, 1) !== '#') { // no configs
            if (!/^\s+/.test(line) && JSON.stringify(hostConfig) !== '{}') {
              this._config.push(hostConfig);
              hostConfig = {};
            }
            const lineData = line.trim().split(/\s+/);
            if (lineData.length > 1) {
              const configName = lineData.shift();
              let configValue = lineData.join(' ');

              if (configName === 'Host') {
                if (configValue.substr(0, 1) === '"') {
                  configValue = configValue.substr(1);
                }
                if (configValue.substr(-1, 1) === '"') {
                  configValue = configValue.substr(0, configValue.length - 1);
                }
              }
              hostConfig[configName] = configValue;
            }
          }
          iterConfig(++cnt, cb);
        } else {
          return cb();
        }
      };
      iterConfig(0, (err2) => {
        callback(err2);
      });
    });
  }

  /**
   * write the ssh config file
   *
   * @param {object[]} hosts
   * @param {function(?error)} callback
   */
  writeConfig(hosts, callback) {
    this.backUpConfig((err) => {
      if (err) {
        return callback(err);
      }
      this._readConfigFile((err2) => {
        if (err2) {
          return callback(err2);
        }
        let yaowstLines = [
          this.CONFIG_BEGIN
        ];
        const iterNewHosts = (cnt, cb) => {
          if (hosts.length > cnt) {
            let hostConfig = hosts[cnt];
            const keys = Object.keys(hostConfig);

            if (hostConfig.Host) {
              yaowstLines.push('Host "' + hostConfig.Host + '"');
              for (let i = 0; i < keys.length; i++) {
                const key = keys[i];
                if (key !== 'Host' && key !== 'X_AutoScalingType') {
                  yaowstLines.push('    ' + key + ' ' + hostConfig[key]);
                }
              }
              yaowstLines.push('');
            }
            setImmediate(() => {
              iterNewHosts(++cnt, cb);
            });
          } else {
            return cb();
          }
        };
        iterNewHosts(0, (err3) => {
          if (err3) {
            return callback(err3);
          }
          let completeConfig = '';
          if (this._stringConfig.ante.length) {
            completeConfig += this._stringConfig.ante.join('\n') + '\n';
          }
          completeConfig += yaowstLines.join('\n') + '\n';
          completeConfig += this.CONFIG_END + '\n';
          if (this._stringConfig.post.length) {
            completeConfig += this._stringConfig.post.join('\n');
          }

          fs.writeFile(this._configFile, completeConfig, {encoding: 'utf8'}, function(err4) {
            callback(err4);
          });
        });
      });
    });
  }

}

module.exports = SshConfig;


lib/config.js
'use strict';

const fs = require('fs');
const path = require('path');
const shell = require('shelljs');
const deepExtend = require('deep-extend');

/**
 * The yaowst config handler
 */
class Config {

  /**
   *
   * @param {string|null} configFile
   */
  constructor(configFile) {
    if (!configFile) {
      configFile = process.env.HOME + '/.yaowst';
    }
    Object.defineProperties(this, {
      _defaultConfig: {
        value: {
          opsWorks: {
            region: 'us-east-1'
          },
          sshConfigFile: {
            file: null,
            saveMode: '24/7'
          },
          sshOptions: {
            StrictHostKeyChecking: 'no',
            UserKnownHostsFile: '/dev/null',
            IdentitiesOnly: 'yes'
          },
          stacks: []
        },
        writable: false
      },
      _configFile: {
        value: configFile,
        writable: false
      },
      _configData: {
        value: null,
        writable: true
      }
    });
  }

  /**
   * read config file callback
   *
   * @callback Config~readConfigFileCallback
   * @param {Error} [error]
   * @param {string} [fileContent]
   */

  /**
   * read the config file
   *
   * @param {Config~readConfigFileCallback} callback
   * @private
   */
  _readConfigFile(callback) {
    fs.stat(this._configFile, (err, stats) => {
      if (err && err.code !== 'ENOENT') {
        return callback(err);
      } else if (err && err.code === 'ENOENT') {
        return callback(new Error('config file dose not exists!'));
      } else if (!stats.isFile()) {
        return callback(new Error('config path is no file!'));
      }

      const fileMode = parseInt(stats.mode.toString(8), 10).toString().substr(-3);
      if (fileMode === '600' || fileMode === '400') {
        fs.readFile(this._configFile, 'utf8', (err2, data) => {
          if (err2) {
            return callback(new Error('Can\'t write the config file' + err2.toString()));
          } else {
            return callback(null, data.trim());
          }
        });
      } else {
        return callback(new Error('The config file "' + this._configFile + '" must have the permission 0400 or 0600'));
      }
    });
  }

  /**
   * write the default config file callback
   *
   * @callback Config~writeDefaulConfigFileCallback
   * @param {Error} [error]
   */

  /**
   * write the default config file
   *
   * @param {boolean} force
   * @param {Config~writeDefaulConfigFileCallback} callback
   */
  writeDefaultFile(force, callback) {
    let defaultConfig = {
      opsWorks: {
        region: 'us-east-1'
      },
      sshOptions: {
        StrictHostKeyChecking: 'no',
        UserKnownHostsFile: '/dev/null',
        IdentitiesOnly: 'yes'
      },
      stacks: []
    };

    // check if directory exists
    const checkConfigDirExists = (cb) => {
      const configDir = path.dirname(this._configFile);
      fs.stat(configDir, (err) => {
        if (err && err.code !== 'ENOENT') {
          return callback(err);
        } else if (err && err.code === 'ENOENT') {
          shell.mkdir('-p', configDir);
        }
        return cb();
      });
    };

    fs.stat(this._configFile, (err) => {
      if (err && err.code !== 'ENOENT') {
        return callback(err);
      } else if (err && err.code === 'ENOENT' || force) {
        checkConfigDirExists((createErr) => {
          if (createErr) {
            return callback(createErr);
          }
          fs.writeFile(this._configFile, JSON.stringify(defaultConfig, null, 2), {mode: '0600'}, (writeErr) => {
            if (writeErr) {
              return callback(new Error('Can\'t write the config file. Error: ' + writeErr.toString()));
            }
            return callback();
          });
        });
      } else {
        return callback(new Error('The config file already exists!'));
      }
    });
  }

  /**
   * get the config data from the config file
   *
   * @param {function(?error, ?object)} callback
   * @return {function(?error, ?object)}
   */
  getConfig(callback) {
    if (!this._configData) {
      this._readConfigFile((err, configString) => {
        if (err) {
          return callback(err);
        }
        try {
          const configJson = JSON.parse(configString);
          const configData = Object.assign({}, this._defaultConfig);

          // extend the default config
          deepExtend(configData, configJson);

          // store the config data
          this._configData = configData;

          return callback(null, configData);
        } catch (err2) {
          return callback(new Error('can\'t parse the config, ' + err2.toString()));
        }
      });
    } else {
      return callback(null, this._configData);
    }
  }

  /**
   * get the config with inherit sshOptions
   *
   * @param {function(?error, ?object)} callback
   */
  getEnhancedConfig(callback) {
    this.getConfig((err, configData) => {
      if (err) {
        return callback(err);
      }
      let stacks = [];
      const iterStacks = (cnt, cb) => {
        if (configData.stacks.length > cnt) {
          let stack = configData.stacks[cnt];
          const baseSshOptions = Object.assign({}, configData.sshOptions);
          const stackSshOptions = stack.sshOptions || {};

          // add ssh options
          stack.sshOptions = Object.assign(baseSshOptions, stackSshOptions);

          // add global opsWorks settings
          if (configData.opsWorks.accessKeyId && !stack.accessKeyId) {
            stack.accessKeyId = configData.opsWorks.accessKeyId;
          }

          if (configData.opsWorks.secretAccessKey && !stack.secretAccessKey) {
            stack.secretAccessKey = configData.opsWorks.secretAccessKey;
          }

          if (configData.opsWorks.region && !stack.region) {
            stack.region = configData.opsWorks.region;
          }

          stacks.push(stack);
          iterStacks(++cnt, cb);
        } else {
          return cb();
        }
      };

      iterStacks(0, function(err2) {
        if (err2) {
          return callback(err2);
        } else {
          configData.stacks = stacks;
          return callback(null, configData);
        }
      });

    });
  }
}

module.exports = Config;
